contrib/babelfishpg_tsql/src/pltsql_ruleutils.c

```diff
@@ -0,0 +1,2441 @@
+ * show it, or 0 to show it only if the constant wouldn't be assumed to be
+ * the right type by default.
+ *
+ * If the Const's collation isn't default for its type, show that too.
+ * We mustn't do this when showtype is -1 (since that means the caller will
+ * print "::typename", and we can't put a COLLATE clause in between).  It's
+ * caller's responsibility that collation isn't missed in such cases.
+ * ----------
+ */
+static void
+get_const_expr(Const *constval, deparse_context *context, int showtype)
+{
+	StringInfo	buf = context->buf;
+	StringInfo	valbuf;
+	Oid			typoutput;
+	bool		typIsVarlena;
+	char	   *extval;
+	bool		needlabel = false;
+
+	valbuf = makeStringInfo();
+
+	if (constval->constisnull)
+	{
+		/*
+		 * Always label the type of a NULL constant to prevent misdecisions
+		 * about type when reparsing.
+		 */
+		appendStringInfoString(valbuf, "NULL");
+		if (showtype >= 0)
+		{
+			appendStringInfo(buf, "CAST(%s AS %s)", valbuf->data,
+							 tsql_format_type_extended(constval->consttype,
+								 		constval->consttypmod,
+										FORMAT_TYPE_TYPEMOD_GIVEN));
+			get_const_collation(constval, context);
+		}
+		else
+		{
+			appendStringInfoString(buf, valbuf->data);
+		}
+		pfree(valbuf->data);
+		return;
+	}
+
+	getTypeOutputInfo(constval->consttype,
+					  &typoutput, &typIsVarlena);
+
+	extval = OidOutputFunctionCall(typoutput, constval->constvalue);
+
+	switch (constval->consttype)
+	{
+		case INT4OID:
+
+			/*
+			 * INT4 can be printed without any decoration, unless it is
+			 * negative; in that case print it as '-nnn'::integer to ensure
+			 * that the output will re-parse as a constant, not as a constant
+			 * plus operator.  In most cases we could get away with printing
+			 * (-nnn) instead, because of the way that gram.y handles negative
+			 * literals; but that doesn't work for INT_MIN, and it doesn't
+			 * seem that much prettier anyway.
+			 */
+			if (extval[0] != '-')
+				appendStringInfoString(valbuf, extval);
+			else
+			{
+				appendStringInfo(valbuf, "'%s'", extval);
+				needlabel = true;	/* we must attach a cast */
+			}
+			break;
+
+		case NUMERICOID:
+
+			/*
+			 * NUMERIC can be printed without quotes if it looks like a float
+			 * constant (not an integer, and not Infinity or NaN) and doesn't
+			 * have a leading sign (for the same reason as for INT4).
+			 */
+			if (isdigit((unsigned char) extval[0]) &&
+				strcspn(extval, "eE.") != strlen(extval))
+			{
+				appendStringInfoString(valbuf, extval);
+			}
+			else
+			{
+				appendStringInfo(valbuf, "'%s'", extval);
+				needlabel = true;	/* we must attach a cast */
+			}
+			break;
+
+		case BOOLOID:
+			if (strcmp(extval, "t") == 0)
+				appendStringInfoString(valbuf, "true");
+			else
+				appendStringInfoString(valbuf, "false");
+			break;
+
+		default:
+			simple_quote_literal(valbuf, extval);
+			break;
+	}
+
+	pfree(extval);
+
+	if (showtype < 0)
+	{
+		appendStringInfoString(buf, valbuf->data);
```

Comment:
We need to free memory of `valbuf->data` before returning

Reply:
It's addressed in the next commit.